File: cli-os/src/policy/pep.js
// Policy Enforcement Point. The enforcement lives here, over the transactional store, NOT in
// the request handler that would benefit from ignoring it. A handler RESERVES budget before a
// provider call (the reservation is the ceiling) and COMMITS the real cost after. A crashed or
// misbehaving handler cannot spend past the cap because the reservation was already committed
// atomically. Budgets are enforced in DOLLARS, per project, per UTC day.
import { tx } from '../state/db.js';
import { nowISO, rid, utcDay } from '../util.js';

function capFor(db, project, defaultCap) {
  const row = db.prepare(`SELECT limit_usd FROM caps WHERE project = ? AND window = 'daily'`).get(project);
  return row ? row.limit_usd : defaultCap;
}

function spendRow(db, project, day) {
  db.prepare(`INSERT OR IGNORE INTO spend(project,day,reserved_usd,committed_usd) VALUES(?,?,0,0)`)
    .run(project, day);
  return db.prepare(`SELECT reserved_usd, committed_usd FROM spend WHERE project = ? AND day = ?`)
    .get(project, day);
}

// Atomically reserve `amountUsd`. Denies if it would breach the daily cap.
export function reserve(db, { project, amountUsd, defaultCap }) {
  // Fail closed on non-finite/negative amounts — they would corrupt spend totals or let a
  // negative "reservation" bypass the cap.
  if (!Number.isFinite(amountUsd) || amountUsd < 0) {
    return { ok: false, reason: 'invalid_amount', requested: amountUsd };
  }
  const day = utcDay();
  return tx(db, () => {
    const cap = capFor(db, project, defaultCap);
    const s = spendRow(db, project, day);
    const inUse = s.reserved_usd + s.committed_usd;
    if (inUse + amountUsd > cap + 1e-9) {
      return { ok: false, reason: 'cost_cap', cap, spent: inUse, requested: amountUsd };
    }
    const id = rid('rsv');
    db.prepare(`INSERT INTO reservations(id,project,day,amount_usd,state,created_at) VALUES(?,?,?,?,'reserved',?)`)
      .run(id, project, day, amountUsd, nowISO());
    db.prepare(`UPDATE spend SET reserved_usd = reserved_usd + ? WHERE project = ? AND day = ?`)
      .run(amountUsd, project, day);
    return { ok: true, reservationId: id, cap, spent: inUse };
  });
}

// Reconcile a reservation with the real cost.
export function commit(db, reservationId, actualUsd) {
  return tx(db, () => {
    const r = db.prepare(`SELECT * FROM reservations WHERE id = ?`).get(reservationId);
    if (!r || r.state !== 'reserved') return false;
    db.prepare(`UPDATE spend SET reserved_usd = reserved_usd - ?, committed_usd = committed_usd + ? WHERE project = ? AND day = ?`)
      .run(r.amount_usd, actualUsd, r.project, r.day);
    db.prepare(`UPDATE reservations SET state = 'committed', amount_usd = ? WHERE id = ?`)
      .run(actualUsd, reservationId);
    return true;
  });
}

export function refund(db, reservationId) {
  return tx(db, () => {
    const r = db.prepare(`SELECT * FROM reservations WHERE id = ?`).get(reservationId);
    if (!r || r.state !== 'reserved') return false;
    db.prepare(`UPDATE spend SET reserved_usd = reserved_usd - ? WHERE project = ? AND day = ?`)
      .run(r.amount_usd, r.project, r.day);
    db.prepare(`UPDATE reservations SET state = 'refunded' WHERE id = ?`).run(reservationId);
    return true;
  });
}

// Reap orphaned reservations. A handler that crashes between reserve() and commit()/refund()
// strands a `reserved` row, which counts against the daily cap until the UTC-day rollover. A
// multi-hop bridge request multiplies that exposure (N reservations per request), so recovery must
// not wait for restart alone. Refund any reservation still `reserved` after maxAgeMs — far longer
// than a legitimate call (retry.maxAttempts * requestTimeoutMs) so an in-flight request is never
// reaped out from under itself. Returns the count reclaimed.
export function reapStaleReservations(db, maxAgeMs) {
  const cutoff = new Date(Date.now() - maxAgeMs).toISOString();
  return tx(db, () => {
    const stale = db.prepare(`SELECT * FROM reservations WHERE state = 'reserved' AND created_at < ?`).all(cutoff);
    for (const r of stale) {
      db.prepare(`UPDATE spend SET reserved_usd = MAX(0, reserved_usd - ?) WHERE project = ? AND day = ?`)
        .run(r.amount_usd, r.project, r.day);
      db.prepare(`UPDATE reservations SET state = 'refunded' WHERE id = ?`).run(r.id);
    }
    return stale.length;
  });
}

export function getSpend(db, project, defaultCap) {
  const day = utcDay();
  const s = spendRow(db, project, day);
  return { day, cap: capFor(db, project, defaultCap), reserved: s.reserved_usd, committed: s.committed_usd };
}

// Concurrency leases (session pool, per-repo memory write). Returns lease id or null if full.
export function acquireLease(db, { project, kind, ttlSec = 300, max = 1 }) {
  return tx(db, () => {
    const now = Date.now();
    // reap expired
    db.prepare(`DELETE FROM leases WHERE expires_at < ?`).run(new Date(now).toISOString());
    const active = db.prepare(`SELECT COUNT(*) c FROM leases WHERE project = ? AND kind = ?`)
      .get(project, kind).c;
    if (active >= max) return null;
    const id = rid('lease');
    db.prepare(`INSERT INTO leases(id,project,kind,created_at,expires_at) VALUES(?,?,?,?,?)`)
      .run(id, project, kind, new Date(now).toISOString(), new Date(now + ttlSec * 1000).toISOString());
    return id;
  });
}

export function releaseLease(db, id) {
  db.prepare(`DELETE FROM leases WHERE id = ?`).run(id);
}

// Destructive-action gate. The chat request path never performs push/merge/deploy/credential
// changes; anything that would must pass an explicit per-action grant (never a blanket start-of-
// run grant). v1 denies by default — there is no destructive surface in the gateway yet.
export function requirePermission(action, grant) {
  if (!grant || grant.action !== action || Date.parse(grant.expires_at || 0) < Date.now()) {
    return { ok: false, reason: 'permission_required', action };
  }
  return { ok: true };
}

File: cli-os/internal/policy/pep.go
// Package policy is the Policy Enforcement Point. Enforcement lives HERE, over the transactional
// store, NOT in the request handler that would benefit from ignoring it. A handler RESERVES budget
// before a provider call (the reservation is the ceiling) and COMMITS the real cost after. Budgets
// are enforced in DOLLARS, per project, per UTC day. Ported from pep.js.
package policy

import (
	"database/sql"
	"math"
	"time"

	"github.com/jackofall1232/l00prite/cli-os/internal/state"
	"github.com/jackofall1232/l00prite/cli-os/internal/util"
)

// ReserveResult mirrors the reserve() return contract.
type ReserveResult struct {
	OK            bool
	ReservationID string
	Reason        string // "invalid_amount" | "cost_cap"
	Cap           float64
	Spent         float64
	Requested     float64
}

// Spend is the current daily spend snapshot.
type Spend struct {
	Day       string
	Cap       float64
	Reserved  float64
	Committed float64
}

func capFor(q state.Querier, project string, defaultCap float64) float64 {
	var limit float64
	err := q.QueryRowContext(state.Ctx(),
		`SELECT limit_usd FROM caps WHERE project = ? AND window = 'daily'`, project).Scan(&limit)
	if err != nil {
		return defaultCap
	}
	return limit
}

func spendRow(q state.Querier, project, day string) (reserved, committed float64, err error) {
	if _, err = q.ExecContext(state.Ctx(),
		`INSERT OR IGNORE INTO spend(project,day,reserved_usd,committed_usd) VALUES(?,?,0,0)`, project, day); err != nil {
		return 0, 0, err
	}
	err = q.QueryRowContext(state.Ctx(),
		`SELECT reserved_usd, committed_usd FROM spend WHERE project = ? AND day = ?`, project, day).
		Scan(&reserved, &committed)
	return reserved, committed, err
}

// Reserve atomically reserves amountUsd, denying if it would breach the daily cap. Fails closed on
// non-finite / negative amounts.
func Reserve(db *sql.DB, project string, amountUsd, defaultCap float64) ReserveResult {
	if math.IsNaN(amountUsd) || math.IsInf(amountUsd, 0) || amountUsd < 0 {
		return ReserveResult{OK: false, Reason: "invalid_amount", Requested: amountUsd}
	}
	day := util.UTCDay()
	res, err := state.Tx(db, func(q state.Querier) (ReserveResult, error) {
		cap := capFor(q, project, defaultCap)
		reserved, committed, serr := spendRow(q, project, day)
		if serr != nil {
			return ReserveResult{}, serr // fail CLOSED: a DB read failure must not proceed on unknown spend
		}
		inUse := reserved + committed
		if inUse+amountUsd > cap+1e-9 {
			return ReserveResult{OK: false, Reason: "cost_cap", Cap: cap, Spent: inUse, Requested: amountUsd}, nil
		}
		id := util.RID("rsv")
		if _, err := q.ExecContext(state.Ctx(),
			`INSERT INTO reservations(id,project,day,amount_usd,state,created_at) VALUES(?,?,?,?,'reserved',?)`,
			id, project, day, amountUsd, util.NowISO()); err != nil {
			return ReserveResult{}, err
		}
		if _, err := q.ExecContext(state.Ctx(),
			`UPDATE spend SET reserved_usd = reserved_usd + ? WHERE project = ? AND day = ?`,
			amountUsd, project, day); err != nil {
			return ReserveResult{}, err
		}
		return ReserveResult{OK: true, ReservationID: id, Cap: cap, Spent: inUse}, nil
	})
	if err != nil {
		// A transaction/DB error denies (fail closed) rather than silently proceeding.
		return ReserveResult{OK: false, Reason: "internal_error"}
	}
	return res
}

// Commit reconciles a reservation with the real cost. Returns true if a reserved row was committed.
func Commit(db *sql.DB, reservationID string, actualUsd float64) bool {
	ok, _ := state.Tx(db, func(q state.Querier) (bool, error) {
		var (
			rProject, rDay, rState string
			rAmount                float64
		)
		err := q.QueryRowContext(state.Ctx(),
			`SELECT project, day, amount_usd, state FROM reservations WHERE id = ?`, reservationID).
			Scan(&rProject, &rDay, &rAmount, &rState)
		if err != nil || rState != "reserved" {
			return false, nil
		}
		if _, err := q.ExecContext(state.Ctx(),
			`UPDATE spend SET reserved_usd = reserved_usd - ?, committed_usd = committed_usd + ? WHERE project = ? AND day = ?`,
			rAmount, actualUsd, rProject, rDay); err != nil {
			return false, err
		}
		if _, err := q.ExecContext(state.Ctx(),
			`UPDATE reservations SET state = 'committed', amount_usd = ? WHERE id = ?`,
			actualUsd, reservationID); err != nil {
			return false, err
		}
		return true, nil
	})
	return ok
}

// Refund releases a still-reserved reservation. Returns true on success.
func Refund(db *sql.DB, reservationID string) bool {
	ok, _ := state.Tx(db, func(q state.Querier) (bool, error) {
		var (
			rProject, rDay, rState string
			rAmount                float64
		)
		err := q.QueryRowContext(state.Ctx(),
			`SELECT project, day, amount_usd, state FROM reservations WHERE id = ?`, reservationID).
			Scan(&rProject, &rDay, &rAmount, &rState)
		if err != nil || rState != "reserved" {
			return false, nil
		}
		if _, err := q.ExecContext(state.Ctx(),
			`UPDATE spend SET reserved_usd = reserved_usd - ? WHERE project = ? AND day = ?`,
			rAmount, rProject, rDay); err != nil {
			return false, err
		}
		if _, err := q.ExecContext(state.Ctx(),
			`UPDATE reservations SET state = 'refunded' WHERE id = ?`, reservationID); err != nil {
			return false, err
		}
		return true, nil
	})
	return ok
}

// ReapStaleReservations refunds any reservation still 'reserved' after maxAgeMs. Returns the count.
func ReapStaleReservations(db *sql.DB, maxAgeMs int64) int {
	cutoff := util.ISOFromTime(time.Now().Add(-time.Duration(maxAgeMs) * time.Millisecond))
	n, _ := state.Tx(db, func(q state.Querier) (int, error) {
		rows, err := q.QueryContext(state.Ctx(),
			`SELECT id, project, day, amount_usd FROM reservations WHERE state = 'reserved' AND created_at < ?`, cutoff)
		if err != nil {
			return 0, err
		}
		type stale struct {
			id, project, day string
			amount           float64
		}
		var list []stale
		for rows.Next() {
			var s stale
			if err := rows.Scan(&s.id, &s.project, &s.day, &s.amount); err != nil {
				continue
			}
			list = append(list, s)
		}
		rows.Close()
		for _, s := range list {
			if _, err := q.ExecContext(state.Ctx(),
				`UPDATE spend SET reserved_usd = MAX(0, reserved_usd - ?) WHERE project = ? AND day = ?`,
				s.amount, s.project, s.day); err != nil {
				return 0, err
			}
			if _, err := q.ExecContext(state.Ctx(),
				`UPDATE reservations SET state = 'refunded' WHERE id = ?`, s.id); err != nil {
				return 0, err
			}
		}
		return len(list), nil
	})
	return n
}

// GetSpend returns the current daily spend snapshot for a project.
func GetSpend(db *sql.DB, project string, defaultCap float64) Spend {
	day := util.UTCDay()
	s, _ := state.Tx(db, func(q state.Querier) (Spend, error) {
		reserved, committed, _ := spendRow(q, project, day) // best-effort read for display
		return Spend{Day: day, Cap: capFor(q, project, defaultCap), Reserved: reserved, Committed: committed}, nil
	})
	return s
}

// AcquireLease grabs a concurrency lease (session pool / per-repo write). Returns ("", false) if full.
func AcquireLease(db *sql.DB, project, kind string, ttlSec, max int) (string, bool) {
	if ttlSec <= 0 {
		ttlSec = 300
	}
	if max <= 0 {
		max = 1
	}
	id, _ := state.Tx(db, func(q state.Querier) (string, error) {
		now := time.Now()
		if _, err := q.ExecContext(state.Ctx(),
			`DELETE FROM leases WHERE expires_at < ?`, util.ISOFromTime(now)); err != nil {
			return "", err
		}
		var active int
		_ = q.QueryRowContext(state.Ctx(),
			`SELECT COUNT(*) FROM leases WHERE project = ? AND kind = ?`, project, kind).Scan(&active)
		if active >= max {
			return "", nil
		}
		leaseID := util.RID("lease")
		if _, err := q.ExecContext(state.Ctx(),
			`INSERT INTO leases(id,project,kind,created_at,expires_at) VALUES(?,?,?,?,?)`,
			leaseID, project, kind, util.ISOFromTime(now), util.ISOFromTime(now.Add(time.Duration(ttlSec)*time.Second))); err != nil {
			return "", err
		}
		return leaseID, nil
	})
	return id, id != ""
}

// ReleaseLease drops a lease by id.
func ReleaseLease(db *sql.DB, id string) {
	_, _ = db.ExecContext(state.Ctx(), `DELETE FROM leases WHERE id = ?`, id)
}

// Grant is a per-action permission grant for the destructive-action gate.
type Grant struct {
	Action    string
	ExpiresAt string
}

// PermissionResult is the outcome of a destructive-action gate check.
type PermissionResult struct {
	OK     bool
	Reason string
	Action string
}

// RequirePermission enforces the destructive-action gate: denies unless a matching, unexpired grant
// is presented. v1 has no destructive surface in the gateway, so this defaults to deny.
func RequirePermission(action string, grant *Grant) PermissionResult {
	if grant == nil || grant.Action != action {
		return PermissionResult{OK: false, Reason: "permission_required", Action: action}
	}
	if t, err := parseISO(grant.ExpiresAt); err != nil || t.Before(time.Now()) {
		return PermissionResult{OK: false, Reason: "permission_required", Action: action}
	}
	return PermissionResult{OK: true}
}

func parseISO(s string) (time.Time, error) {
	if t, err := time.Parse(time.RFC3339Nano, s); err == nil {
		return t, nil
	}
	return time.Parse("2006-01-02T15:04:05.000Z07:00", s)
}
